### PR TITLE
Add draggy number input

### PR DESCRIPTION
**Pull request by g10ria**
Component takes a speed option, set to an integer for integer input and a real number for float input

### CODE REVIEW

**Comment by 132ikl** [comment]
Can you make these changes:
* Fix the arrows
* Disable dragging when in manual edit mode
* Hide the default number-type input arrows on Firefox (https://ikl.sh/taCd)
* Separate speed and type control (just add a "float" parameter with boolean value)
* Rename speed to step for clarity

**Review by 132ikl — CHANGES_REQUESTED** [review]
Can you make these changes:
* Fix the arrows
* Disable dragging when in manual edit mode
* Hide the default number-type input arrows on Firefox (https://ikl.sh/taCd)
* Separate speed and type control (just add a "float" parameter with boolean value)
* Rename speed to step for clarity

**Comment by g10ria** [comment]
What do you mean by fix the arrows?

**Comment by 132ikl** [comment]
The increment/decrement arrows on the left/right don't work for me on Chrome or Firefox.

**Comment by g10ria** [comment]
Oh huhh that's weird. Ok I'll look into it

**Review by 132ikl — APPROVED** [review]

**Review by 132ikl — CHANGES_REQUESTED** [review]
Can you make sure that the pipeline update event isn't sent until the user releases dragging? Currently each step sends an update to the backend.

**Review by 132ikl — APPROVED** [review]

File: src/components/baklava-options-plugin/DragFloatOption.vue [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * Parameters:
+ * defaultValue - default starting value for number input
+ * step - how 'fast' the value is decreased/increased when sliding or using the arrow keys
+ *      defaults to 1
+ * floatPrecision - precision to which the float should be shown
+ *      defaults to 3
+ */
+export default class DragIntegerOption extends DragNumberOption {
```
Comment (carbotaniuman) [inline src/components/baklava-options-plugin/DragFloatOption.vue:11]: I think you need an `@Component` here.

File: src/components/baklava-options-plugin/DragIntegerOption.vue [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<script lang="ts">
+import DragNumberOption from "./DragNumberOption.vue";
+
+/**
+ * Parameters:
+ * defaultValue - default starting value for number input
+ * step - how 'fast' the value is decreased/increased when sliding or using the arrow keys
+ *      defaults to 1
+ */
+export default class DragIntegerOption extends DragNumberOption {
```
Comment (carbotaniuman) [inline src/components/baklava-options-plugin/DragIntegerOption.vue:11]: I think you need an `@Component` here.

File: src/components/baklava-options-plugin/DragNumberOption.vue [hunk 3]
```diff
@@ -16,46 +14,75 @@
...
+    const step = this.option.step || 1;
+    this.displayValue += e.movementX * step;
+  }
+
+  created() {
+    if (this.option.defaultValue) this.displayValue = this.option.defaultValue;
+  }
+
+  startDrag() {
+    const moveListener = this.handleMouseMove.bind(this);
```
Comment (carbotaniuman) [inline src/components/baklava-options-plugin/DragNumberOption.vue]: Why not an arrow function?

File: src/components/baklava-options-plugin/DragNumberOption.vue [hunk 4]
```diff
@@ -16,46 +14,75 @@
...
+  }
 
-  // @Model('change', { type: Number }) readonly clicking!: boolean
+  handleMouseMove(e: MouseEvent) {
+    const step = this.option.step || 1;
+    this.displayValue += e.movementX * step;
+  }
+
+  created() {
+    if (this.option.defaultValue) this.displayValue = this.option.defaultValue;
```
Comment (carbotaniuman) [inline src/components/baklava-options-plugin/DragNumberOption.vue]: this.value (the prop) gets set with the right value at some point in time. Idk if it's at  `mounted()` or `created()`, but Baklava already has that built in, so let's use it.

**Review by carbotaniuman — CHANGES_REQUESTED** [review]

**Comment by 132ikl** [comment]
Can you do the following in order to fix up the history:
1. Reset to 96a88f7
2. Pull with rebase option
3. Cherry-pick f7d5ad3
4. Force push

**Review by 132ikl — CHANGES_REQUESTED** [review]
Can you change it to make sure that if the min/max is not set then it is not limited?

**Review by 132ikl — APPROVED** [review]

File: src/components/baklava-options-plugin/DragIntegerOption.vue [hunk 5]
```diff
@@ -0,0 +1,22 @@
+<script lang="ts">
+import DragNumberOption from "./DragNumberOption.vue";
+
+/**
+ * Parameters:
+ * defaultValue - default starting value for number input
+ * step - how 'fast' the value is decreased/increased when sliding or using the arrow keys
+ *      defaults to 1
+ * min - minimum value
+ *      defaults to -100
```
Comment (carbotaniuman) [inline src/components/baklava-options-plugin/DragIntegerOption.vue]: Outdated.

File: src/components/baklava-options-plugin/DragFloatOption.vue [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ * Parameters:
+ * defaultValue - default starting value for number input
+ * step - how 'fast' the value is decreased/increased when sliding or using the arrow keys
+ *      defaults to 1
+ * floatPrecision - precision to which the float should be shown
+ *      defaults to 3
+ * min - minimum value
+ *      defaults to -100
+ * max - maximum value
```
Comment (carbotaniuman) [inline src/components/baklava-options-plugin/DragFloatOption.vue]: Outdated.

File: src/components/baklava-options-plugin/DragNumberOption.vue [hunk 7]
```diff
@@ -16,46 +14,84 @@
...
-  // @Model('change', { type: Number }) readonly clicking!: boolean
+    const handleMouseUp = () => {
+      document.removeEventListener("mousemove", moveListener);
+      document.removeEventListener("mouseup", handleMouseUp);
+
+      this.setVal(this.displayValue);
+
+      // if no change from the mouse events, it was probably a click, so enter edit mode
+      if (originalVal === this.displayValue) this.enterEditMode();
+      if (this instanceof DragIntegerOption) this.tempValue = this.displayValue.toFixed(0);
```
Comment (carbotaniuman) [inline src/components/baklava-options-plugin/DragNumberOption.vue:76]: tempValue never gets used anywhere? is this meant to be displayValue?

File: src/components/baklava-options-plugin/DragNumberOption.vue [hunk 8]
```diff
@@ -16,46 +14,84 @@
...
+  }
+
+  handleMouseMove(e: MouseEvent) {
+    const step = this.option.step || 1;
+    this.displayValue = this.clamp(this.displayValue + e.movementX * step);
+  }
+
+  clamp(num: number) {
+    let ret = num;
+    if (this.option.min) ret = Math.max(this.option.min, ret);
```
Comment (carbotaniuman) [inline src/components/baklava-options-plugin/DragNumberOption.vue]: test for `=== undefined`, these prevent 0 mins and maxes

File: src/components/baklava-options-plugin/DragNumberOption.vue [hunk 9]
```diff
@@ -16,46 +14,84 @@
...
+    if (this.option.max) ret = Math.min(this.option.max, ret);
```
Comment (carbotaniuman) [inline src/components/baklava-options-plugin/DragNumberOption.vue]: test for `=== undefined`, these prevent 0 mins and maxes

**Review by carbotaniuman — CHANGES_REQUESTED** [review]

File: src/components/baklava-options-plugin/DragNumberOption.vue (hunk 7 above)
Comment (carbotaniuman) [reply src/components/baklava-options-plugin/DragNumberOption.vue:76]: Looks like it's inherited? Just verify its correct.

**Review by carbotaniuman — APPROVED** [review]
LGTM

**Pull request merged**